src/computerAlgorithms.js

```diff
@@ -0,0 +1,202 @@
+const lines = [
+    [0, 1, 2],
+    [3, 4, 5],
+    [6, 7, 8],
+    [0, 3, 6],
+    [1, 4, 7],
+    [2, 5, 8],
+    [0, 4, 8],
+    [2, 4, 6],
+];
+const diagonalLines = lines.slice(6);
+const straightLines = lines.slice(0,6);
+
+function firstStepAlgorithm(squares, playerGoesFirst){
+  if (playerGoesFirst){
+    if (!squares[4]) {
+      return 4;
+    } else {
+      return getRandomAnglePosition(squares);
+    }
+  } else {
+    const firstStepInCenter = Math.round(Math.random());
+
+    if (firstStepInCenter) {
+      return 4;
+    } else {
+      return getRandomAnglePosition(squares);
+    }
+  }
+}
+
+function secondStepAlgorithm(squares, playerGoesFirst){
+  if (playerGoesFirst) {
+    if (squares[4] === 'O') {
+      const checkOppositeDiagonals = (squares[0] === 'X' && squares[0] === squares[8]) || (squares[2] === 'X' && squares[2] === squares[6]);
+      const checkXOnSiblingSides = (squares[1] === 'X' && (squares[1] === squares[3] || squares[1] === squares[5])) || (squares[7] === 'X' && (squares[7] === squares[3] || squares[7] === squares[5]));
+      const checkXOnSides = findMarkOn(squares, 'X', 'side');
+      const checkXOnAngles = findMarkOn(squares, 'X', 'angle');
+
+      if(checkOppositeDiagonals) {
+        return getRandomSidePosition(squares);
+      } else if (checkXOnSiblingSides || (checkXOnSides && checkXOnAngles)) {
+        return getPositionNearXOnAngle(squares);
+      }
+    }
+
+    return getRandomAnglePosition(squares);
+  }
+
+  if (squares[4] !== 'X') {
+    if (squares[4] !== 'O') {
+      let [isZeroOnAngle,] = findMarkOn(squares, 'O', 'angle');
+
+      if (isZeroOnAngle) {
+        return getRandomAnglePosition(squares);
+      } else {
+        return getAnglePositionOnFriendlyLine(squares);
+      }
+    } else {
+      for(let i = 0; i < diagonalLines.length; i ++) {
+        const [a, , c] = diagonalLines[i];
+
+        if (squares[a] === 'X' && !squares[c]) {
+          return c;
+        } else if (squares[c] === 'X' && !squares[a]) {
+          return a;
+        }
+      }
+    }
+  }
+
+  if (squares[4] === 'X') {
+    return getRandomAnglePosition(squares);
+  }
+}
+
+function findMarkOn(squares, mark, where) {
+  let i = where === 'angle' ? 0 : 1;
+  let isMarkOn = false;
+  let markPosition = null;
+
+  while(i < squares.length) {
+    if (squares[i] === mark && i !== 4) {
```

Comment:
you use 4 many time, you should save 4 in variable with a name, because it looks like a magic number and we don't use them

Reply:
Thanks, I fixed it